FLV: Drop packet if header flag is not matched. v5.0.109

winlinvip opened the pull request:
1. Ignore audo or video packets if FLV header disable it.
2. Reset has_audio if got some video frames but no audio frames.
3. Reset has_video if got some audio frames but no video frames.
4. Note that audio/video frames are not sequence header.

xiaozhihong commented on trunk/src/app/srs_app_http_stream.cpp:
```diff
@@ -307,18 +307,31 @@ srs_error_t SrsFlvStreamEncoder::write_tags(SrsSharedPtrMessage** msgs, int coun
 
     // For https://github.com/ossrs/srs/issues/939
```
Is it possible that calling this function only dumps to a single package (audio, video) and causes misjudgment?

`TRANS_BY_GPT3`

winlinvip replied:
If there is a delay in the flow, such as when converting RTC to RTMP, it sometimes happens. That is definitely a misjudgment, and there is no way to guess it accurately.

If the gop cache is disabled, it still cannot be guessed accurately. In this case, there is only a Sequence Header, and in some exceptional scenarios, there may be AV Sequence Header even though there are no audio packets. In such cases, it is also inaccurate.

Therefore, this issue still requires support for parameters or configuration settings in order to be fully resolved. [Link to the issue](https://github.com/ossrs/srs/issues/939#issuecomment-1348740526)

`TRANS_BY_GPT3`

xiaozhihong commented on trunk/src/kernel/srs_kernel_flv.cpp:
```diff
@@ -481,17 +490,19 @@ srs_error_t SrsFlvTransmuxer::write_tags(SrsSharedPtrMessage** msgs, int count)
+            nb_iovss_cache = nn_might_iovss;
+            iovss = iovss_cache = new iovec[nn_might_iovss];
+        }
+    } while (false);
```
This do while seems to have no effect, can we just remove it?

`TRANS_BY_GPT3`

winlinvip replied:
The main purpose is to hide the scope of `nn_might_iovss`. It would cause issues if this variable is used later on.

The variable that should be used later on is `nn_real_iovss`.

`TRANS_BY_GPT3`

xiaozhihong commented on trunk/src/kernel/srs_kernel_flv.cpp:
```diff
@@ -444,6 +449,8 @@ srs_error_t SrsFlvTransmuxer::write_metadata(char type, char* data, int size)
 {
     srs_error_t err = srs_success;
+
+    if (!has_audio_) return err;
```
The single line is concise, but the code style doesn't match well with the overall SRS.

`TRANS_BY_GPT3`

winlinvip replied:
If it can be done in one line, it is preferable not to make it three lines. Having fewer lines is more friendly for reading code.

For example, this function:

```cpp
srs_error_t SrsFlvTransmuxer::write_tags(SrsSharedPtrMessage** msgs, int count) {
    for (int i = 0; i < count; i++) {
        if (msg->is_audio()) {
            if (!has_audio_) continue; // Ignore audio packets if no audio stream.
            cache_audio(msg->timestamp, msg->payload, msg->size, cache);
        } else if (msg->is_video()) {
            if (!has_video_) continue; // Ignore video packets if no video stream.
            cache_video(msg->timestamp, msg->payload, msg->size, cache);
```

If it is changed to multiple lines, it would be like this:

```cpp
srs_error_t SrsFlvTransmuxer::write_tags(SrsSharedPtrMessage** msgs, int count) {
    for (int i = 0; i < count; i++) {
        if (msg->is_audio()) {
            // Ignore audio packets if no audio stream.
            if (!has_audio_) {
                continue;
            }
            cache_audio(msg->timestamp, msg->payload, msg->size, cache);
        } else if (msg->is_video()) {
            // Ignore video packets if no video stream.
            if (!has_video_) {
                continue;
            }
            cache_video(msg->timestamp, msg->payload, msg->size, cache);
```

In this exceptional case, it is preferable to write it in one line. However, I don't think there is any problem with continuing to write it as before.

`TRANS_BY_GPT3`

winlinvip commented on trunk/src/kernel/srs_kernel_flv.cpp:
```diff
@@ -509,38 +520,41 @@ srs_error_t SrsFlvTransmuxer::write_tags(SrsSharedPtrMessage** msgs, int count)
-        // cache all flv header.
+        // Cache FLV packet header.
         if (msg->is_audio()) {
+            if (!has_audio_) continue; // Ignore audio packets if no audio stream.
```
If the header does not have an audio or video tag, discard the packet. It feels like this should have an additional configuration, as sometimes the packet should not be discarded.

`TRANS_BY_GPT3`

winlinvip committed:
FLV: Drop packet if header flag is not matched. v5.0.109

1. Ignore audo or video packets if FLV header disable it.
2. Run: Add regression test config and run for IDEA.
3. Test: Refine regression test to allow no audio/video for FLV
4. Config: Whether drop packet if not match header.

winlinvip committed:
Merge branch '5.0release' into feature/flv-drop-av
